### ISSUE TITLE
Action should be performed if it is shown as successful

### ISSUE THREAD

**Issue by sguptatw** [getgauge/taiko#81]
**Expected behavior**
Action should be performed if it is shown as successful

**Actual behavior**
The REPL shows that the action is successful, but the element is not clicked

**Steps to replicate**
* Run the script with taiko
```
const { openBrowser, goto, below, image, click } = require('taiko');

(async () => {
   try {
       await openBrowser();
       await goto("amazon.in");
       await click(image(below("Today's Deals")));
       await click("Deal of the Day");
   } catch (e) {
       console.error(e);
   } finally {
       closeBrowser();
   }
})();
```
> The last action is shown as successful, even though it is not performed.

**Version**
```
commit - ffddb5823fa8f966bea64377065f5649bf626f96
Fri Jul 27 09:35:17 2018 +0530
```

**Comment by Apoorva-GA** [comment]
Looks like the amazon website has changed and I don't see `Deal of the day` link.
Could you please recheck this?

**Comment by sguptatw** [comment]
Please find another script to replicate the issue. 
> In the script below, `Book Tickets` is commented. This has to be manually clicked.

```
const { browser, openBrowser, goto, click,link,image } = require('taiko');

(async () => {
    try {
        await openBrowser({headless:false});
        await goto("https://in.bookmyshow.com/bengaluru");

        await click("Bengaluru");
        await write('NCR', into(await inputField({"placeholder":"Search for your city"})));
        await click("NCR");
        await click("View All");
        await click(link("English"))
        await click(image(below(link("Now Showing"))))
        //await click(link("Book Tickets"))
        await waitFor("5000")

        await click(link({"class":"__showtime-link"}))
        await click("Accept")        
        await click(link(toRightOf("D")))
        await click("Select Seats")
        await waitFor("5000")
    } catch (e) {
        console.error(e);
    } finally {
        await closeBrowser();
    }
})();
```
> After step `click(link(toRightOf("D")))` is shown as successful, even though there is no such element to click.

**Comment by Apoorva-GA** [comment]
`await click(link(toRightOf("D")))` is shown successful because the page below the pop has a link to the right of D. So the element actually exists in the above case.

To actually view the action try `await highlight(link(toRightOf("B")));` in this case the highlited link can be seen.

But the question is, whats the expected behaviour?

**Comment by zabil** [comment]
> To actually view the action try await highlight(link(toRightOf("B"))); in this case the highlited link can be seen.

> But the question is, whats the expected behaviour?

In many cases when pop up is displayed it masks elements behind the page. Visually there's no way to interact with elements. How does the pop up look on this page? (screenshot helps!)

**Comment by Apoorva-GA** [comment]
It looks like this.

<img width="1184" alt="screen shot 2018-08-14 at 11 52 47" src="https://user-images.githubusercontent.com/8462534/44075161-40f1e284-9fb9-11e8-87d3-7229f95fc687.png">

**Comment by zabil** [comment]
Ok. This makes sense. 

Ideally or by default it should limit it to the popup area. Is there a way to identify the background masking?

**Comment by Apoorva-GA** [comment]
Not sure. Will have to see if CRI gives any such masking.

**Comment by zabil** [comment]
In the page above, the modal popup is rendered as follows

```
<div class="modal">
	<!-- Global overlay -->
	<div class="__overlay" style="display: block;"></div>
```

The `display: block` display a full height `div` hiding the other elements. So inspect the DOM, to check if the element is visible before clicking it.

**Comment by NivedhaSenthil** [comment]
We do filter visible elements using offsetParent, seems it does not cover all the edge cases. Might have to try something like this https://stackoverflow.com/a/33456469/3336401

**Comment by NivedhaSenthil** [comment]
Fix should be available in fd86932.

**Comment by sguptatw** [comment]
In the following case, click on `Sign in` is shown as successful. However this action is not performed
```
const { openBrowser, goto, scrollTo, click, button, link } = require('taiko');

(async () => {
    try {
        await openBrowser({headless:false});
        await goto("http://demos.jquerymobile.com/1.4.5/popup/");
        await scrollTo("sign in");
        await click("Sign in");
        await click(button("Sign in"));
        await click(link("Sign in"));
    } catch (e) {
        console.error(e);
    } finally {
        closeBrowser();
    }
})();
```

**Comment by NivedhaSenthil** [comment]
Happening due to https://github.com/getgauge/taiko/issues/131

**Comment by Apoorva-GA** [comment]
Should be fixed as part of 888638d.
Please verify.
